File: Daily Questions/2416-sum-of-prefix-scores-of-strings/2416-sum-of-prefix-scores-of-strings.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define ALPHABET_SIZE 26

/**
 * @brief TrieNode structure representing a node in the Trie.
 * Each TrieNode contains:
 * - children: An array of TrieNode pointers representing the 26 lowercase English letters.
 * - count: The number of words that pass through this node.
 */
typedef struct TrieNode {
    struct TrieNode *children[ALPHABET_SIZE];  // Array for child nodes
    int count;  // Number of words passing through this node
} TrieNode;

/**
 * @brief Initializes a new TrieNode.
 *
 * @return TrieNode*: Pointer to the newly created TrieNode.
 */
TrieNode* createNode() {
    TrieNode *node = (TrieNode*)malloc(sizeof(TrieNode));
    node->count = 0;  // Initialize count to 0
    for (int i = 0; i < ALPHABET_SIZE; i++) {
        node->children[i] = NULL;  // Initialize all child nodes to NULL
    }
    return node;
}

/**
 * @brief Inserts a word into the Trie and increments the count of each prefix.
 *
 * @param root TrieNode*: The root of the Trie.
 * @param word char*: The word to be inserted.
 */
void insert(TrieNode *root, const char *word) {
    TrieNode *node = root;
    while (*word) {
        int index = *word - 'a';  // Map 'a' to 0, 'b' to 1, ..., 'z' to 25
        if (node->children[index] == NULL) {
            node->children[index] = createNode();  // Create a new TrieNode if it doesn't exist
        }
        node = node->children[index];
        node->count += 1;  // Increment the count at this node
        word++;
    }
}

/**
 * @brief Calculates the sum of prefix scores for a given word.
 *
 * @param root TrieNode*: The root of the Trie.
 * @param word char*: The word for which to calculate the prefix score.
 * @return int: The sum of prefix scores for the word.
 */
int calculatePrefixScore(TrieNode *root, const char *word) {
    TrieNode *node = root;
    int total_score = 0;  // Initialize total score to 0
    while (*word) {
        int index = *word - 'a';  // Map 'a' to 0, ..., 'z' to 25
        node = node->children[index];
        total_score += node->count;  // Add the count at this node to the total score
        word++;
    }
    return total_score;
}

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
/**
 * @brief Calculates the sum of prefix scores for each word in the list.
 * 
 * @param words char**: List of words.
 * @param wordsSize int: The number of words in the list.
 * @param returnSize int*: Pointer to store the size of the result array.
 * 
 * @return int*: The array of sums of prefix scores for each word.
 *         The returned array is dynamically allocated and must be freed by the caller.
 */
int* sumPrefixScores(char** words, int wordsSize, int* returnSize) {
    // Create a new Trie and initialize the root node
    TrieNode *root = createNode();
    
    // Insert each word into the Trie
    for (int i = 0; i < wordsSize; i++) {
        insert(root, words[i]);  // Insert each word into the Trie
    }
    
    // Allocate memory for the result array
    int *result = (int*)malloc(wordsSize * sizeof(int));
    
    // Calculate the prefix score for each word and store in the result array
    for (int i = 0; i < wordsSize; i++) {
        result[i] = calculatePrefixScore(root, words[i]);
    }
    
    // Set the return size to the number of words
    *returnSize = wordsSize;
    
    // Return the result array
    return result;
}

File: 2416-sum-of-prefix-scores-of-strings/2416-sum-of-prefix-scores-of-strings.py
class TrieNode:
    """
    TrieNode represents a single node in the Trie.
    Each node contains:
    - children: A dictionary to store child nodes corresponding to each character.
    - count: An integer that keeps track of how many words pass through this node.
    """
    def __init__(self):
        self.children = {}  # Dictionary to store child TrieNodes for each character.
        self.count = 0  # Count of words passing through this node.

class Trie:
    """
    Trie is a data structure used for efficiently storing strings and their prefixes.
    It supports the following operations:
    - insert: To insert a word into the Trie.
    - calculate_prefix_score: To calculate the total score of all prefixes of a given word.
    """
    def __init__(self):
        # Initialize the root of the Trie
        self.root = TrieNode()

    def insert(self, word):
        """
        Inserts a word into the Trie. As the word is inserted, it increments the count for each prefix.

        Parameters:
        word (str): The word to insert into the Trie.
        """
        node = self.root
        for char in word:
            # If the character is not in the current node's children, create a new node.
            if char not in node.children:
                node.children[char] = TrieNode()
            # Move to the next node in the Trie
            node = node.children[char]
            # Increment the count to indicate this prefix is part of another word.
            node.count += 1

    def calculate_prefix_score(self, word):
        """
        Calculates the sum of scores for every prefix of a given word.
        The score of each prefix is the number of words that have that prefix.

        Parameters:
        word (str): The word for which to calculate the prefix score.

        Returns:
        int: The sum of the scores of all prefixes of the word.
        """
        node = self.root
        total_score = 0  # Initialize total score to 0
        for char in word:
            # Move to the next node for the current character
            node = node.children[char]
            # Add the count at the current node to the total score
            total_score += node.count
        return total_score

class Solution:
    def sumPrefixScores(self, words: List[str]) -> List[int]:
        """
        Calculates the sum of prefix scores for each word in the input list.

        Parameters:
        words (List[str]): List of words to compute prefix scores for.

        Returns:
        List[int]: A list of integers where each element is the sum of prefix scores of a corresponding word.
        """
        # Initialize the Trie
        trie = Trie()

        # Insert all words into the Trie
        for word in words:
            trie.insert(word)

        # Calculate prefix scores for each word
        result = []
        for word in words:
            # For each word, compute its prefix score using the Trie
            result.append(trie.calculate_prefix_score(word))

        return result